src/content/learn/typescript.md

```diff
@@ -0,0 +1,433 @@
+
+
+## Hooks {/*typing-hooks*/}
+
+The type definitions from `@types/react-dom` include types for the built-in hooks, so you can use them in your components without any additional setup. They are built to take into account the code you write in your component, so you will get inferred types a lot of the time and ideally do not need to handle the minutiae of providing the types. However, we can look at a few examples of how to provide types for hooks.
+
+### `useState` {/*typing-usestate*/}
+
+The [`useState` hook](/reference/react/useReducer) will re-use the value passed in as the initial state to determine what the type of the value should be. For example:
+
+```ts
+const [enabled, setEnabled] = useState(false);
+```
+
+Will assign the type of `boolean` to `enabled`, and `setEnabled` will be a function accepting either a `boolean` argument, or a function that returns a `boolean`. If you want to explicitly provide a type for the state, you can do so by providing a type argument to the `useState` call:
+
+```ts 
+const [enabled, setEnabled] = useState<boolean>(false);
+```
+
+This isn't very useful in this case, but a common case where you may want to provide a type is when you have a union type. For example, `status` here can be one of a few different strings:
+
+```ts
+type Status = "idle" | "loading" | "success" | "error";
+
+const [status, setStatus] = useState<Status>("idle");
+```
+
+Or, as recommended in [Principles for structuring state](/learn/choosing-the-state-structure#principles-for-structuring-state), you can group related state as an object and describe the different possibilities via object types:
+
+```ts
+type RequestState =
+  | { status: 'idle' }
+  | { status: 'loading' }
+  | { status: 'success', data: any }
+  | { status: 'error', error: Error };
+
+const [requestState, setRequestState] = useState<RequestState>({ status: 'idle' });
+```
+
+### `useReducer` {/*typing-usereducer*/}
+
+The [`useReducer` hook](/reference/react/useReducer) is a more complex hook that takes a reducer function and an initial state. The types for the reducer function are inferred from the initial state. You can optionally provide a type argument to the `useReducer` call to provide a type for the state, but it is often better to set the type on the initial state instead:
+
+<Sandpack>
+
+```tsx App.tsx active
+import {useReducer} from 'react';
+
+interface State {
+   count: number 
+};
+
+type CounterAction =
+  | { type: "reset" }
+  | { type: "setCount"; value: State["count"] }
+
+const initialState: State = { count: 0 };
```

Comment:
Is this a good place to showcase `satisfies`? Or would that be distracting from the example?

Reply:
Satisfies is an interesting case, but not what we want here, for example:

```ts
type State = {
  something?: string
  count: number
}

const initial: State = { count: 0 }
const initial2 = { count: 0} satisfies State
```

Would mean that the reducer does not know about `something?: string` on `initial2` but does in `initial`. Because it doesnt modify the type exposed in initial - just the raw literal as the type.

Reply:
Yeah makes sense. I think there are cases where `satisfies` is useful but it probably would warrant a whole separate example of it's own.

Reply:
It could actually be safest to show a horrible
```js
type State = {
  something?: string
  count: number
}

const initial = { count: 0 } satisfies State as State
```
as from my experience, if someone were to move the `initial` declaration into the component before the `useReducer` call, TypeScript code flow analysis might otherwise kick in and the reducer state ends up typed as `{ count: 0 }`, since that it "the only value it could have had on that code path".